File: java/UniquePaths.java
/*
 * Filename: UniquePaths.java
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Mon 18 May 2015 07:03:53 PM EDT
 */

public class Solution {
	public int uniquePaths(int m, int n) {
		if (m == 1 || n == 1) {
			return 1;
		}

		int[] A = new int[n];
		for (int i = 1; i <= n; i++){
			A[i-1] = i;
		}

		for (int i = 2; i < m; i++) {
			for (int j = 1; j < n; j++) {
				A[j] += A[j-1];
			}
		}

		return A[n-1];  
	}
}

File: cpp/unique_paths.cc
/*
 * Filename: unique_paths.cc
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Fri 31 Jul 2015 04:33:54 PM EDT
 */

#include <iostream>
using namespace std;

class Solution {
	public:
		int uniquePaths(int m, int n) {
			if (!m || !n) {
				return 0;
			}

			int dp[n];
			for (size_t j = 0; j < n; j++) {
				dp[j] = 1;
			}

			for (size_t i = 1; i < m; i++) {
				for (size_t j = 1; j < n; j++) {
					dp[j] += dp[j-1];
				}
			}

			return dp[n-1];
		}
};
